File: problems/quick-chats/solutions/quickchats.scala
import java.util.Scanner

// Adapted from https://rosettacode.org/wiki/Dijkstra%27s_algorithm#Scala

object Dijkstra {
  type Path[Key] = (Int, List[Key])
 
  def Dijkstra[Key](lookup: Map[Key, List[(Int, Key)]], fringe: List[Path[Key]], dest: Key, visited: Set[Key]): Path[Key] = fringe match {
    case (dist, path) :: fringe_rest => path match {case key :: path_rest =>
      if (key == dest) (dist, path.reverse)
      else {
        val paths = lookup(key).flatMap {case (d, key) => if (!visited.contains(key)) List((dist + d, key :: path)) else Nil}
        val sorted_fringe = (paths ++ fringe_rest).sortWith {case ((d1, _), (d2, _)) => d1 < d2}
        Dijkstra(lookup, sorted_fringe, dest, visited + key)
      }
    }
    case Nil => (0, List())
  }
}

object Main {
  def main(args: Array[String]): Unit = {
    val SENTINEL = 1000000000

    val input = new Scanner(System.in)
    val n = input.nextLine().toInt
    val names = input.nextLine().split(' ')
    val m = input.nextLine().toInt

    val lookup = scala.collection.mutable.Map[String, List[(Int, String)]]()
    (0 to n-1) foreach (i => {
      val adj: List[(Int, String)] = names.map { n =>
        val cost = if (names(i) == n) 0 else SENTINEL
        (cost, n)
      }.toList

      lookup.put(names(i), adj)
    })

    (1 to m) foreach (i => {
      val line = input.nextLine().split(' ')
      val s = line(0)
      val d = line(1)
      val c = line(2).toInt
      val t = line(3).toInt
      val adj: List[(Int, String)] = lookup.get(s).get
      val newAdj = adj :+ (c + t, d)
      lookup.put(s, newAdj)
    })

    val lastLine = input.nextLine().split(' ')
    val source = lastLine(0)
    val dest = lastLine(1)

    val minCost = Dijkstra.Dijkstra[String](lookup.toMap, List((0, List(source))), dest, Set())._1
    println(if (minCost == SENTINEL) -1 else minCost)
  }
}

File: problems/quick-chats/solutions/quickchats.py
#!/usr/bin/python3

import heapq

SENTINEL = 1E9

if __name__ == '__main__':
	n = int(input())
	names = []
	times = {}

	for _ in range(n):
		name, time = input().split()
		names.append(name)
		times[name] = int(time)

	m = int(input())
	graph = {}
	for n in names:
		graph[n] = {}

	for _ in range(m):
		[s, d, c] = input().split()
		graph[s][d] = int(c) + times[d] if s != d else 0

	[source, dest] = input().split()

	dist = {}
	q = []

	for n in names:
		dist[n] = 0 if n == source else SENTINEL
		heapq.heappush(q, (dist[n], n))

	while len(q) > 0:
		(_, u) = heapq.heappop(q)
		for v in graph[u]:
			alt = dist[u] + graph[u][v]
			if alt < dist[v]:
				dist[v] = alt
				heapq.heappush(q, (dist[v], v))

	print(dist[dest] if dist[dest] != SENTINEL else -1)
